src/main/java/com/spotify/dns/AbstractChangeNotifier.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2012-2015 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.spotify.dns;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A helper for implementing the {@link ChangeNotifier} interface.
+ */
+abstract class AbstractChangeNotifier<T> implements ChangeNotifier<T> {
+
+  private final AtomicReference<Listener<T>> listenerRef = new AtomicReference<Listener<T>>();
+
+  @Override
+  public void setListener(final Listener<T> listener, final boolean fire) {
+    checkNotNull(listener, "listener");
+
+    if (!listenerRef.compareAndSet(null, listener)) {
+      throw new IllegalStateException("Listener already set!");
+    }
+
+    if (fire) {
+      fireEndpointsUpdated(newChangeNotification(current(), Collections.<T>emptySet()));
+    }
+  }
+
+  @Override
+  public final void close() {
+    listenerRef.set(null);
+
+    closeImplementation();
+  }
+
+  protected abstract void closeImplementation();
+
+  protected void fireEndpointsUpdated(ChangeNotification<T> changeNotification) {
+    checkNotNull(changeNotification, "changeNotification");
+
+    final Listener<T> listener = listenerRef.get();
+    if (listener != null) {
+      listener.endpointsChanged(changeNotification);
+    }
+  }
+
+  protected ChangeNotification<T> newChangeNotification(Set<T> current, Set<T> previous) {
+    checkNotNull(current, "current");
+    checkNotNull(previous, "previous");
+
+    return new ChangeNotificationImpl(current, previous);
+  }
+
+  private class ChangeNotificationImpl implements ChangeNotification<T> {
```

Comment:
should be private static class, I guess? It doesn't seem to need a reference to the outer class.